<a href="https://colab.research.google.com/github/sdiaz8/sdiaz8/blob/main/Groupby_with_tips.csv.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving tips.csv to tips (4).csv


In [ ]:
import io
df = pd.read_csv(io.BytesIO(uploaded['tips.csv']))

#Changing sex, smoker, and day to meaningful values

In [ ]:
df['sex'] = df['sex'].replace([0,1],['male','female'])
df['smoker'] = df['smoker'].replace([0,1],['NO','YES'])
df['day'] = df['day'].replace([4,5,6,7],['Thursday','Friday','Saturday','Sunday'])

In [ ]:
df.head()

total_bill   tip     sex smoker     day    time  size
0       16.99  1.01    male     NO  Sunday  Dinner     2
1       10.34  1.66  female     NO  Sunday  Dinner     3
2       21.01  3.50  female     NO  Sunday  Dinner     3
3       23.68  3.31  female     NO  Sunday  Dinner     2
4       24.59  3.61    male     NO  Sunday  Dinner     4

#Question 1

In [ ]:
filt1 = df["total_bill"] > 40
filt2 = df["size"] <= 4

df[filt1 & filt2]

total_bill    tip     sex smoker       day    time  size
59        48.27   6.73  female     NO  Saturday  Dinner     4
95        40.17   4.73  female    YES    Friday  Dinner     4
102       44.30   2.50    male    YES  Saturday  Dinner     3
170       50.81  10.00  female    YES  Saturday  Dinner     3
182       45.35   3.50  female    YES    Sunday  Dinner     3
184       40.55   3.00  female    YES    Sunday  Dinner     2
197       43.11   5.00    male    YES  Thursday   Lunch     4
212       48.33   9.00  female     NO  Saturday  Dinner     4

In [ ]:
print("There are", len(df[filt1 & filt2]), "orders that have a bill greater than $40 and a party size of less than 4.")

There are 8 orders that have a bill greater than $40 and a party size of less than 4.


#Question 2

In [ ]:
df['sex'].value_counts()

female    157
male       87
Name: sex, dtype: int64

Women pay the bill more often than Men.

#Question 3

In [ ]:
df.groupby(['day','time']).agg({'total_bill':'mean'}).sort_values('total_bill',ascending=False).round(2)

total_bill
day      time              
Sunday   Dinner       21.41
Saturday Dinner       20.44
Friday   Dinner       19.66
Thursday Dinner       18.78
         Lunch        17.66
Friday   Lunch        12.85

The time and day that has the highest average bill is Sunday Dinner with $21.41. 

Friday Lunch has the lowest average bill at $12.85.

#Question 4

In [ ]:
df.groupby('day').agg({'size':'mean'}).sort_values('size', ascending=False).round(2)

size
day           
Sunday    2.84
Saturday  2.52
Thursday  2.45
Friday    2.11

On average, the waiter serves the largest parties on Sundays.

#Question 5

In [ ]:
df.groupby(['day','time','size']).agg({'total_bill':'mean'}).sort_values(['day','time']).round(2)

total_bill
day      time   size            
Friday   Dinner 2          17.80
                4          40.17
         Lunch  1           8.58
                2          13.07
                3          15.98
Saturday Dinner 1           5.16
                2          16.84
                3          25.51
                4          29.88
                5          28.15
Sunday   Dinner 2          17.56
                3          22.18
                4          26.69
                5          27.00
                6          48.17
Thursday Dinner 2          18.78
         Lunch  1          10.07
                2          15.08
                3          19.16
                4          29.95
                5          41.19
                6          30.38

#Question 6

In [ ]:
df['tip_rate'] = (df.tip / df.total_bill).round(4)
df.sort_values('tip_rate',ascending=True)

total_bill   tip     sex smoker       day    time  size  tip_rate
237       32.83  1.17  female    YES  Saturday  Dinner     2    0.0356
102       44.30  2.50    male    YES  Saturday  Dinner     3    0.0564
57        26.41  1.50    male     NO  Saturday  Dinner     2    0.0568
0         16.99  1.01    male     NO    Sunday  Dinner     2    0.0594
187       30.46  2.00  female    YES    Sunday  Dinner     5    0.0657
..          ...   ...     ...    ...       ...     ...   ...       ...
183       23.17  6.50  female    YES    Sunday  Dinner     4    0.2805
232       11.61  3.39  female     NO  Saturday  Dinner     2    0.2920
67         3.07  1.00    male    YES  Saturday  Dinner     1    0.3257
178        9.60  4.00    male    YES    Sunday  Dinner     2    0.4167
172        7.25  5.15  female    YES    Sunday  Dinner     2    0.7103

[244 rows x 8 columns]

The lowest tip rate was 3.56%, which was paid by a female smoker.